### PR TITLE
🌱  Add lifecycle hook types

### PR DESCRIPTION
**Pull request by killianmuldoon**
Co-authored-by: ykakarap <ykakaraparth@vmware.com>
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add types for the lifecycle hooks defined in the lifecycle hook proposal.

Part of #6330 
Implements the types from proposal #5141 

/area runtime-sdk

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
There's an outstanding question in https://github.com/kubernetes-sigs/cluster-api/issues/6506
About whether we should compose the types by creating shared structs used throughout the top-level types.

I'm happy to do that in this PR if that's the direction we're looking to go.

**Comment by sbueringer** [comment]
> There's an outstanding question in #6506 About whether we should compose the types by creating shared structs used throughout the top-level types.
> 
> I'm happy to do that in this PR if that's the direction we're looking to go.

As discussed yesterday, let's finalize #6506 and then depending on timing potentially update this PR to embed shared structs in the response types.

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Status of the call. One of "Success" or "Failure".
+	Status ResponseStatus `json:"status"`
+
+	// RetryAfterSeconds when set to a non-zero signifies that the hook
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:43]:
```suggestion
	// RetryAfterSeconds when set to a non-zero value signifies that the hook
```
nit, same below

**Comment by sbueringer** [comment]
/hold cancel
#5141 has been merged

**Comment by killianmuldoon** [comment]
/hold cancel
This is unblocked by adding the CommonResponse type from #6499

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 2]
```diff
@@ -16,6 +16,12 @@ limitations under the License.
 
 package v1alpha1
 
+// CommonResponse is the data structure common to all response types.
+type CommonResponse struct {
+	Message string         `json:"message"`
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:21]: Might make sense to sync it with the version after my last nit on the other PR

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:41]:
```suggestion
	CommonResponse `json:",inline"`
```
same below

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go (hunk 1 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:43]: *bump* (just to make sure it's not forgotten)

**Comment by sbueringer** [comment]
Thank you!
/lgtm

/assign @fabriziopandini

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+}
+
+// AfterControlPlaneInitialized is the runtime hook that will be called after the control plane is available for the first time.
+func AfterControlPlaneInitialized(*AfterControlPlaneInitializedRequest, *AfterControlPlaneInitializedResponse) {
+}
+
+// BeforeClusterUpgradeRequest is the request of the hook.
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:73]:
```suggestion
// BeforeClusterUpgradeRequest is the request of the BeforeClusterUpgrade hook.
```
nit

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains a Status and Message field common to all response types.
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:40]:
```suggestion
	// CommonResponse contains a Status and Message fields common to all response types.
```
nit

(same comment on all other uses)

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	// RetryAfterSeconds when set to a non-zero value signifies that the hook
+	// needs to be retried at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+}
+
+// BeforeClusterUpgrade is the runtime hook that will be called after a cluster.spec.version is upgraded and
+// before the updated version is propagated to the underlying objects.
+func BeforeClusterUpgrade(*BeforeClusterUpgradeRequest, *BeforeClusterUpgradeResponse) {}
+
+// AfterControlPlaneUpgradeRequest is the request of the hook.
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:104]:
```suggestion
// AfterControlPlaneUpgradeRequest is the request of the AfterControlPlaneUpgrade hook.
```
nit (for consistency with other comments)

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+
+	// RetryAfterSeconds when set to a non-zero value signifies that the hook
+	// needs to be retried at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+}
+
+// AfterControlPlaneUpgrade is the runtime hook after the control plane is successfully upgraded to the target
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:129]:
```suggestion
// AfterControlPlaneUpgrade is the runtime hook called after the control plane is successfully upgraded to the target
```

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+// kubernetes version and before the target version is propagated to the workload machines.
+func AfterControlPlaneUpgrade(*AfterControlPlaneUpgradeRequest, *AfterControlPlaneUpgradeResponse) {}
+
+// AfterClusterUpgradeRequest is the request of the hook.
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:133]:
```suggestion
// AfterClusterUpgradeRequest is the request of the AfterClusterUpgrade hook.
```

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+}
+
+// AfterClusterUpgrade is the runtime hook that is called after all of the cluster is updated
+// to the target kubernetes version.
+func AfterClusterUpgrade(*AfterClusterUpgradeRequest, *AfterClusterUpgradeResponse) {}
+
+// BeforeClusterDeleteRequest is the request of the hook.
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:158]:
```suggestion
// BeforeClusterDeleteRequest is the request of the BeforeClusterDelete hook.
```

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+	catalogBuilder.RegisterHook(AfterControlPlaneInitialized, &runtimecatalog.HookMeta{
+		Tags:        []string{"Lifecycle Hooks"},
+		Summary:     "Called after the Control Plane is available for the first time",
+		Description: "This non-blocking hook is called after the ControlPlane for the Cluster is marked as available for the first time",
+	})
+
+	catalogBuilder.RegisterHook(BeforeClusterUpgrade, &runtimecatalog.HookMeta{
+		Tags:        []string{"Lifecycle Hooks"},
+		Summary:     "Called before the Cluster begins upgrade",
+		Description: "This hook is called after the Cluster object has been updated with a new spec.topology.version by the user, and immediately before the new version is propagated to the Control Plane",
```
Comment (ykakarap) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:200]:
```suggestion
		Description: "This blocking hook is called after the Cluster object has been updated with a new spec.topology.version by the user, and immediately before the new version is propagated to the Control Plane",
```

**Review by ykakarap — COMMENTED** [review]
Overall LGTM.

Just a few minor nits. Not blocking.

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go (hunk 5 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:40]: Kilians suggestion sounds like the best option to me, but I trust the native speaker on that one :)

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains Status and Message fields common to all response types.
+	CommonResponse `json:",inline"`
+}
+
+// AfterControlPlaneInitialized is the runtime hook that will be called after the control plane is available for the first time.
+func AfterControlPlaneInitialized(*AfterControlPlaneInitializedRequest, *AfterControlPlaneInitializedResponse) {
```
Comment (chrischdi) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:70]: Should we also document here **why** the `AfterControlPlaneInitialized` hook is non-blocking? I did search for the reason in https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220330-topology-mutation-hook.md but did not find any.

**Pull request merged**